transports/tls: Add `libp2p-tls` as per spec

thomaseizinger opened the pull request:
<!-- Please write a summary of your changes and why you made them.-->

## Description

This code is mostly extracted from #2289. I've tagged everyone who made contributions to that PR as co-authors (cc @kpp @dvc94ch @elenaf9 @mxinden) in the first commit.

On top of that, this also adds an implementation of the `Upgrade{Inbound,Outbound}` traits so that TLS can be used with other transports. Finally, this PR also makes an effort towards https://github.com/libp2p/specs/issues/459 by including a few static tests (cc @marten-seemann).



## Links to any relevant issues

- https://github.com/libp2p/specs/issues/459
- https://github.com/libp2p/rust-libp2p/pull/2289
- https://github.com/libp2p/rust-libp2p/pull/1175
- https://github.com/ChainSafe/forest/issues/974 (cc @austinabell)
- Fixes https://github.com/libp2p/rust-libp2p/issues/1109.

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

## Open tasks

- [ ] Reference https://github.com/DemiMarie as co-author: `Co-authored-by:  Demi Marie Obenour <demiobenour@gmail.com>`
- [ ] Reference https://github.com/tomaka as co-author: `Co-authored-by: Pierre Krieger <pierre.krieger1708@gmail.com>`

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Add skeleton for libp2p-tls

This code is extracted from https://github.com/libp2p/rust-libp2p/pull/2289.

Co-authored-by: David Craven <david@craven.ch>
Co-authored-by: Roman Proskuryakov <r.proskuryakoff@gmail.com>
Co-authored-by: Max Inden <mail@max-inden.de
Co-authored-by: Elena Frank <elena.frank@protonmail.com>

thomaseizinger committed:
Minor reformatting

thomaseizinger committed:
Fix compile errors and remove `Error` type

thomaseizinger committed:
Remove unnecessary `pub(crate)`

thomaseizinger committed:
Implement `Upgrade` traits to allow use of TLS as authentication

thomaseizinger committed:
Add changelog files

thomaseizinger committed:
Fix docs

thomaseizinger committed:
Minor doc and name adjustments

thomaseizinger committed:
Verify p2p extension as part of parsing certificate

We shouldn't rely on our callers to call `verify` here.

thomaseizinger committed:
Fix compile error

thomaseizinger committed:
Extract private fn that allows to control the cert keypair

This will be useful for testing.

thomaseizinger committed:
fixup! Verify p2p extension as part of parsing certificate

thomaseizinger committed:
Add static test vectors

thomaseizinger committed:
Re-order items by importance

thomaseizinger committed:
Re-order functions in impl block and make private where possible

thomaseizinger committed:
Tidy up manifest

thomaseizinger committed:
Sort hex dependencies

thomaseizinger committed:
Fix clippy

thomaseizinger committed:
Remove `Stream` type

We are already leaking the `rustls` dependency to our clients
through the `make_{server,client}_config` functions which have to
be public for QUIC. Might as well reduce boilerplate and not wrap
`TlsStream`.

thomaseizinger committed:
Fmt

thomaseizinger committed:
Exclude `libp2p-tls` from wasm build

thomaseizinger committed:
Add license to libp2p-tls

thomaseizinger committed:
Don't leak `rcgen` dependency in public API

thomaseizinger committed:
Don't leak `webpki` dependency

thomaseizinger committed:
Add note about exposed dependency to manifest

thomaseizinger committed:
Upgraxe x509-parser to 0.14

Co-authored-by: Roman Proskuryakov <r.proskuryakoff@gmail.com>

thomaseizinger committed:
Remove comment

kpp commented on transports/tls/src/certificate.rs:
```diff
@@ -0,0 +1,547 @@
+///
+/// For this to succeed, the certificate must contain the specified extension and the signature must
+/// match the embedded public key.
+pub fn parse(der_input: &[u8]) -> Result<P2pCertificate, ParseError> {
```
parse_verify?

thomaseizinger replied:
I purposely I named this one `parse` and the other one `parse_unverified` because I think verification / validation should be part of parsing to make good use of the type system. Also see [this blogpost](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).

kpp commented on transports/tls/src/verifier.rs:
```diff
@@ -0,0 +1,217 @@
+}
+
+impl ServerCertVerifier for Libp2pCertificateVerifier {
+    fn verify_server_cert(
```
So we can dial with a PK and without. If we dial with a PK then we should get a cert.extension.public_key and check whether they are equal. This code was mistakenly removed in 3b1ff348999da1543.

cc @marten-seemann

kpp replied:
From specs:

```
The public host key allows the peer to calculate the peer ID of the peer it is connecting to. Clients MUST verify that the peer ID derived from the certificate matches the peer ID they intended to connect to, and MUST abort the connection if there is a mismatch.
```

marten-seemann replied:
> So we can dial with a PK and without. If we dial with a PK then we should get a cert.extension.public_key and check whether they are equal. This code was mistakenly removed in https://github.com/libp2p/rust-libp2p/commit/3b1ff348999da1543b7c394f4db9f7364e061508.
>
> cc @marten-seemann

Not sure I understand the question, can you please clarify? PK = public key? private key?
The libp2p extension needs to be present in the certificate in any case.

kpp replied:
PK = public key.

> The libp2p extension needs to be present in the certificate in any case.

Yes but if we dial with a known PK we MUST verify that the PK in the extension equals to the PK we are dialing to. Is it correct?

marten-seemann replied:
That's correct. Although I'd expect the more common case to be that you know the Peer ID in advance, not the public key (for peer IDs where the key is not inlined).

thomaseizinger replied:
Verification that we connected to the intended peer (`/p2p`) happens in the `Swarm`, independent of the given `Transport` in `rust-libp2p`. I think we are covering this requirement for any `Transport` as long as people are using the `Swarm` abstraction[^1].


[^1]: Most users are as that is what a lot of functionality is built on top of but in theory, you could use it without `libp2p-swarm`. I know of one project that does and there, the check is implemented manually.

kpp commented on transports/tls/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
```
Should exclude publishing test_assets

thomaseizinger replied:
Yep, good catch, thank you!

kpp commented on transports/tls/src/certificate.rs:
```diff
@@ -43,9 +43,7 @@ static P2P_SIGNATURE_ALGORITHM: &rcgen::SignatureAlgorithm = &rcgen::PKCS_ECDSA_
-pub fn generate(
-    identity_keypair: &identity::Keypair,
-) -> Result<rcgen::Certificate, rcgen::RcgenError> {
+pub fn generate(identity_keypair: &identity::Keypair) -> Result<rcgen::Certificate, GenError> {
```
Still `rcgen::Certificate` leaks?

thomaseizinger replied:
You are right, well spotted!

We can make this one `pub(crate)` though because it is only used by internal APIs.

thomaseizinger replied:
Resolved in [`e1f34ae` (#2945)](https://github.com/libp2p/rust-libp2p/pull/2945/commits/e1f34aeafb9a543465e52d6bfffdfeb620cd9843).

kpp commented on transports/tls/tests/smoke.rs:
```diff
@@ -0,0 +1,73 @@
+            };
+        }
+    };
+    swarm2.dial(listen_address).unwrap();
```
This one dials something like `/memory/2698462084451899396` (without PK). It would be nice to have two more tests: with a correct explicit PK and with incorrect PK.

thomaseizinger replied:
Verifying whether we dialed the correct peer in case the multiaddr has a peer ID inlined happens on the `Swarm` level in `rust-libp2p` and is thus transport independent. I don't think we need to add a test for this!

marten-seemann replied:
That works, but it’s not a very nice behavior. Certificates should be verified during the handshake, as this allows you to abort the connection attempt during TLS handshake.
Verifying the peer ID after the handshake leads to the successful establishment of a connection on the server side, just to have that connection killed 1 RTT later. Even worse, while aborting during the handshake lets the peer know why the handshake failed (as this is communicated by a TLS alert), aborting after handshake completion leaves the peer guessing what went wrong.

thomaseizinger replied:
Yeah that makes sense. It is a historical decision that would require a fair bit of work to change. I'll open an issue about it but would prefer to not block this PR on it.

kpp replied:
> but would prefer to not block this PR on it.

And we will merge master and use a not full implementation of the tls spec in 2 PRs. I can do that if you like.

kpp replied:
https://github.com/libp2p/rust-libp2p/pull/2947

thomaseizinger committed:
Exclude test assets from publishing

thomaseizinger committed:
Move errors further down

These are not interesting enough to be as prominent.

thomaseizinger committed:
Inline `_generate` function

This is a left-over from an earlier iteration where we thought
that we need to control the keypair for deterministic tests.

thomaseizinger committed:
Extract `make_libp2p_extension` function

thomaseizinger committed:
Use `rustls` types in signatures

This actually hides the `rcgen` dependency from our users. We already
leak `rustls` which is basically unavoidable for as long as we want
to expose the `make_{server,client}_config` functions so it seems
adequate to just use those types in the `certificate` module too.

mxinden commented on transports/tls/src/certificate.rs:
```diff
@@ -0,0 +1,552 @@
+/// Internal function that only parses but does not verify the certificate.
+///
+/// Useful for testing but unsuitable for production.
+fn parse_unverified(der_input: &[u8]) -> Result<P2pCertificate, webpki::Error> {
```
```suggestion
#[cfg(test)]
fn parse_unverified(der_input: &[u8]) -> Result<P2pCertificate, webpki::Error> {
```

Not an optimization, but rather a safety measure to prevent accidental use. What do you think?

thomaseizinger replied:
Can't do that because it is used by the `parse` function to avoid code duplication.

It is private so I don't think we need to be worried about accidential use.

mxinden commented on transports/tls/tests/smoke.rs:
```diff
@@ -0,0 +1,73 @@
+use futures::{future, StreamExt};
```
Please add a license header here :innocent: and in all other files.

mxinden reviewed (commented):
Wonderful to see this happening. Thanks @thomaseizinger and everyone involved.

I have not yet taken a deeper look.

@thomaseizinger have you been able to test this against a go-libp2p node? E.g. against one of `/dnsaddr/bootstrap.libp2p.io/`, e.g. via https://github.com/mxinden/libp2p-lookup?

kpp committed:
libp2p-tls: Allow to specify remote peer ID we intend to connect to (#2947)

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

thomaseizinger committed:
Point to correct version of `libp2p-tls`

thomaseizinger committed:
Polish public API and expose via root crate

thomaseizinger committed:
Add license headers

thomaseizinger committed:
Fix compile error in test

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

mxinden commented on transports/tls/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+libp2p-core = { version = "0.37.0", path = "../../core" }
+rcgen = "0.9.2"
+ring = "0.16.20"
+rustls = { version = "0.20.2", default-features = false, features = ["dangerous_configuration"] } # Exposed dependency! Bump major version on any change.
```
Can you document why it is safe here to enable the `dangerous_configuration` feature?

thomaseizinger replied:
Sure thing!

mxinden reviewed (commented):
This looks good to me. Thanks @thomaseizinger for the follow-ups. Thanks @kpp for https://github.com/libp2p/rust-libp2p/pull/2945/commits/79a464a780e67f04f952e90440cfd6e69a9b184e.

Given that the logic here is critical to get right, I would appreciate additional eyes. @marten-seemann do you have some time to give this a review? Happy to help with any Rust related questions.

marten-seemann commented on transports/tls/src/verifier.rs:
```diff
@@ -0,0 +1,242 @@
+        // The libp2p handshake uses TLS 1.3 (and higher).
+        // Endpoints MUST NOT negotiate lower TLS versions.
+        Err(rustls::Error::PeerIncompatibleError(
+            "Only TLS 1.3 certificates are supported".to_string(),
```
Why is this check done when verifying the certificate? The TLS version is negotiated before certificates are exchanged.

I also believe that the term "TLS 1.3 certificate" is misleading. Certificates are not bound to the TLS version.

kpp replied:
> The TLS version is negotiated before certificates are exchanged.

That's right and this is how it is done right now. So this is a no-op just in case.

thomaseizinger replied:
> Why is this check done when verifying the certificate? The TLS version is negotiated before certificates are exchanged.

The TLS library at hand supports multiple TLS versions but they are all combined into one interface. Meaning at compile-time, we don't know which TLS version will be used and thus we need to satisfy this interface and provide an implementation for `verify_tls12_signature`. This function will should never be called because we disable anything other than 1.3 at configuration time. Due to TLS' flexibility, this is hard to express statically so I assume the authors of the TLS library didn't bother in modelling this using the type system.

> I also believe that the term "TLS 1.3 certificate" is misleading. Certificates are not bound to the TLS version.

I'll change the error message.

marten-seemann commented on transports/tls/src/certificate.rs:
```diff
@@ -0,0 +1,552 @@
+
+    /// Return the signature scheme corresponding to [`AlgorithmIdentifier`]s
+    /// of `subject_pki` and `signature_algorithm`
+    /// according to `<https://tools.ietf.org/id/draft-ietf-tls-tls13-21.html#rfc.section.4.2.3>`.
```
This should link to RFC 8446.

marten-seemann reviewed (approved):
Not sure my review actually adds a lot of value, since my Rust knowledge is very limited.

I was surprised to see so much code related to different signature algorithms, is there really no way to have the TLS library handle that for you?

thomaseizinger committed:
Make it more explicit which protocol versions we support

thomaseizinger committed:
Move config of supported cipher suits into `verifier` module

thomaseizinger committed:
Fix clippy lints

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

thomaseizinger committed:
Remove unnecessary `pub`

thomaseizinger committed:
Clarify why we need `dangerous_configuration`

thomaseizinger committed:
Fix link to signature scheme

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

thomaseizinger committed:
Adapt to new `swarm` interface

bkchr commented on transports/tls/src/certificate.rs:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 Parity Technologies (UK) Ltd.
+// Copyright 2022 Protocol Labs.
```
Just removing the copyright doesn't look right to me. I'm not a lawyer, but the original copyright should stay?

thomaseizinger replied:
That was an oversight, thanks for pointing that out!

thomaseizinger committed:
Restore original copy-right

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

mxinden commented on transports/tls/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+[package]
+name = "libp2p-tls"
+version = "0.1.0"
```
```suggestion
version = "0.1.0-alpha"
```

To the best of my knowledge, the code below has not run in any production systems.

To communicate expectations, I suggest releasing this as an alpha first.

See also same suggestion on QUIC pull request https://github.com/libp2p/rust-libp2p/pull/2289#discussion_r983395958

thomaseizinger replied:
On it!

Most users depend on `libp2p` though so they are not going to see this version. IMO we are effectively removing that "alpha" label again through the re-export.

Do you think we should perhaps not re-export the TLS implementation yet so users have to depend on it manually?

mxinden replied:
> Most users depend on `libp2p` though so they are not going to see this version. IMO we are effectively removing that "alpha" label again through the re-export.

That is true. Though I would explicitly call it out in the changelog.

> Do you think we should perhaps not re-export the TLS implementation yet so users have to depend on it manually?

I don't have an opinion here. I think we should be consistent with `libp2p-quic`. Chatted with @elenaf9 quickly. She prefers not removing `libp2p-quic` from `libp2p`.

elenaf9 replied:
What do you think about annotating the re-export with the [`unstable` attribute macro](https://rustc-dev-guide.rust-lang.org/stability.html#unstable) @thomaseizinger ?

thomaseizinger replied:
> What do you think about annotating the re-export with the [`unstable` attribute macro](https://rustc-dev-guide.rust-lang.org/stability.html#unstable) @thomaseizinger ?

That is specific to the Rust compiler from what I know, I don't think you can use that actually?

elenaf9 replied:
Ah yes you are right! Don't have a strong on opinion on whether to re-export tls and quic or not. I did not consider that when re-exporting we "hide" the alpha-flag, so maybe not re-exporting is the best solution after all.

thomaseizinger replied:
I think not exporting it would be the better solution but I don't want to block this PR on it. We can always change that until the next release.

mxinden reviewed (approved):
This is good to merge from my end. Please go ahead @thomaseizinger unless @elenaf9 has any objections.

I am in favor of adding the above list of co-authors to the squash commit.

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

thomaseizinger committed:
Set `libp2p-tls` to `0.1.0-alpha`

thomaseizinger committed:
Fix clippy error in latest rustls release

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

thomaseizinger committed:
Merge branch 'master' into libp2p-tls

elenaf9 commented on transports/tls/src/lib.rs:
```diff
@@ -0,0 +1,78 @@
+//! TLS configuration based on libp2p TLS specs.
+//!
+//! See <https://github.com/libp2p/specs/blob/master/tls/tls.md>.
+
```
```suggestion

#![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]

```

thomaseizinger committed:
Use `doc_auto_cfg` for libp2p-tls

thomaseizinger committed:
Merge branch 'libp2p-tls' of github.com:libp2p/rust-libp2p into libp2p-tls
